Use newgistics_id for syncing Spree::Variant

narma opened the pull request.

jmataya commented on app/async/workers/inventory_puller.rb:
```diff
@@ -24,10 +24,14 @@ def update_inventory(newgistics_stock_items)
+        unless variant then
+          variant = variants.find { |v| v.sku == newgistic_stock_item['sku'] }
+          variant.update_columns(newgistics_id: newgistic_stock_item['id'].to_i) if variant
+        end
```
Why not combine this into a simple test?

```ruby
variant = variants.find do |v| 
  v.newgistics_id == newgistics_stock_item["id"] || v.sku == newgistics_stock_tiem["sku"]
end
variant.update_columns(newgistics_id: newgistics_stock_item["id"].to_i) if variant
```

Doing both checks in the context of a single lookup seems like it should be more performant.

narma replied:
Agree, your variant looks simpler.

jmataya commented on app/async/workers/products_puller.rb:
```diff
@@ -68,7 +67,11 @@ def save_products(products)
+          if spree_variant.nil?
+            spree_variant = Spree::Variant.find_by_sku(product['sku'])
+            spree_variant.update_columns(newgistics_id: product['id'].to_i) if spree_variant
+          end
```
Same comment as above. Seems like you could do this in one `find` lookup.

narma replied:
Well, there we use not `find` lookup, but doing `find_by`. So, there we can use `where` method for generate sql with `OR`, because `find` search only be one parameter.

jmataya commented:
Looks good!

Just check out the couple comments that I made and merge master.

narma commented:
@jmataya I've updated PR, but can you recheck my second approach? I use `Arel` features for generate sql query with `OR`.

narma commented:
I see very weird errors in tests: https://buildkite.com/foxcommerce/spree-newgistics/builds/86#ee18337c-b478-4621-ad28-bb152ea3307f . Looks into it

narma commented:
@jmataya I've changed logic in the last commit. I switched to previous variant where firstly we search by `newgistics_id`and then if we haven't found anything, we search by `sku`. Because main point of this PR was in the search of Variant by id from NG, and when we do `||` in lookup or `OR` in sql we break this logic.

jmataya commented:
I think this new logic makes sense. Barring any comments from @Axblade, it looks good to me.

Axblade commented:
What's the reason to update `newgistics_id` in `ProductsPuller` [here](https://github.com/FoxComm/spree_newgistics/pull/36/files#diff-14c81bae0b0b8205fa6a2f3c5f321a30R77) and `InventoryPuller` [here](https://github.com/FoxComm/spree_newgistics/pull/36/files#diff-a7ace61727eea537446fce39a1388ab4R35)? I think it should be only in `ProductsPuller`, as its work is to update products. `InventoryPuller` should update inventory only.

Except this comment, it looks good to me.

narma commented:
It makes sense, also I've removed redundant conidition in `ProductsPuller`

Axblade commented on app/async/workers/products_puller.rb:
```diff
@@ -68,13 +67,15 @@ def save_products(products)
           item_category_id = item_category_id_from(product['supplier'], spree_categories)
+          shipping_category_id = hazardous_category_id if product['customFields'] && product['customFields']['hazMatClass'].eql?('ORM-D')
 
-          shipping_category_id = hazardous_category_id if product['customFields'] && (product['customFields']['hazMatClass'].eql?('ORM-D') || product['customFields']['HazMatClass'].eql?('ORM-D'))
```
I'm not sure that it's redundant. Looks like there can be categories with capitalize version and lowercase one.

narma replied:
Oh, good catch, I'll revert my change

jmataya commented:
Good catch on the HazMat capitalization, @Axblade!
